lib/fs/promises.js

```diff
@@ -147,15 +147,18 @@ async function readFileHandle(filehandle, options) {
 
   const chunks = [];
   const chunkSize = Math.min(size, 16384);
-  const buf = Buffer.alloc(chunkSize);
   let totalRead = 0;
+  let endOfFile = false;
   do {
+    const buf = Buffer.alloc(chunkSize);
     const { bytesRead, buffer } =
-      await read(filehandle, buf, 0, buf.length);
-    totalRead = bytesRead;
-    if (totalRead > 0)
-      chunks.push(buffer.slice(0, totalRead));
-  } while (totalRead === chunkSize);
+      await read(filehandle, buf, 0, chunkSize, totalRead);
+    totalRead += bytesRead;
+    if (bytesRead !== chunkSize)
+      endOfFile = true;
```

Comment:
nit: replace these two lines with `endOfFile = bytesRead !== chunkSize;`

Reply:
Isn't that very less efficient to set the variable at each iteration?

Reply:
About 5% slower according to the benchmarks I just ran, so "very less" is a bit overstating, still it's not nothing.

Reply:
Interesting, I assumed it would be faster due to lack of branching (at least it would be in a language like C). Oh well.